### PR TITLE
[deckhouse-controller] create separate controller for DeckhouseRelease

### LINKED ISSUE #8374: [deckhouse-controller] create separate controller for DeckhouseRelease
### Preflight Checklist

- [x] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

Don't block the update process if something is broken in deckhouse modules logic

### Proposed Solution

1. Create the controller [here](https://github.com/deckhouse/deckhouse/tree/main/deckhouse-controller/pkg/controller)
2. Move hooks logic to this controller
  - https://github.com/deckhouse/deckhouse/blob/main/modules/002-deckhouse/hooks/check_deckhouse_release.go (use goroutine for checking the release image)
  - https://github.com/deckhouse/deckhouse/blob/main/modules/002-deckhouse/hooks/cleanup_deckhouse_releases.go (some scheduled goroutine for this (once per day and onStartup would be enough)
  - https://github.com/deckhouse/deckhouse/blob/main/modules/002-deckhouse/hooks/update_deckhouse_image.go (same as ModuleRelease controller)
3. Fetch Deckhouse config values as [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/controller.go#L252)

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by libmonsoon-dev**
## Description
Deckhouse update logic has been moved to a separate controller

## Why do we need it, and what problem does it solve?
Closes #8374

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: Deckhouse update logic has been moved to a separate controller
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Comment by libmonsoon-dev** [comment]
/e2e/abort openstack;Standard;containerd;1.29 9607120284 9607120284-1-con-1-29 /sys/deckhouse-oss/install:pr8667

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+		DisruptionApprovalMode: "Auto",
+		NotificationConfig:     new(updater.NotificationConfig),
+	}
+
+	// default case in template
+	if data.ReleaseChannel == "Unknown" {
+		data.ReleaseChannel = ""
+	}
+
+	clusterBootstrapped, ok := secret.Data["clusterIsBootstrapped"]
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/controller.go:260]: NIT: Don't you want to unmarshal all needed secret data to some struct, and not to ready it directly?
You have to bother about some details, like quotes

File: deckhouse-controller/pkg/controller/deckhouse-release/cleanup_deckhouse_release.go [hunk 2]
```diff
@@ -14,66 +14,57 @@ See the License for the specific language governing permissions and
...
+	err := r.client.List(ctx, &releases)
+	if err != nil {
+		return fmt.Errorf("get deckhouse releases: %w", err)
+	}
 
-	releases := make([]*d8updater.DeckhouseRelease, 0, len(snap))
-	for _, sn := range snap {
-		releases = append(releases, sn.(*d8updater.DeckhouseRelease))
+	pointerReleases := make([]*v1alpha1.DeckhouseRelease, 0, len(releases.Items))
+	for _, r := range releases.Items {
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/cleanup_deckhouse_release.go:54]: NIT: shadowing. Reconciler has the same variable

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+		windows = r.updatePolicy.Get().Update.Windows
+	}
+
+	deckhouseUpdater.ApplyPredictedRelease(windows)
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
+}
+
+func (r *deckhouseReleaseReconciler) getDeckhousePods(ctx context.Context) ([]corev1.Pod, error) {
+	var pods corev1.PodList
+	err := r.client.List(
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/controller.go:201]: Guess you have to add them [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/controller.go#L131)

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+
+	return &secret, nil
+}
+
+// TODO: get from release object
+func (r *deckhouseReleaseReconciler) getReleaseData(ctx context.Context) (updater.DeckhouseReleaseData, error) {
+	var cm corev1.ConfigMap
+
+	key := types.NamespacedName{Namespace: "d8-system", Name: "d8-release-data"}
+	err := r.client.Get(ctx, key, &cm)
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/controller.go:413]: Also have to add [here](https://github.com/deckhouse/deckhouse/blob/main/deckhouse-controller/pkg/controller/controller.go#L131)

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+const defaultCheckInterval = 15 * time.Second
+
+type deckhouseReleaseReconciler struct {
+	client        client.Client
+	dc            dependency.Container
+	logger        logger.Logger
+	moduleManager moduleManager
+
+	updatePolicy            *v1alpha1.ModuleUpdatePolicySpecContainer
+	releaseVersionImageHash *container[string]
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/controller.go:60]: Do we need this container here? It's only used in the synchronous function

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+	if data.ReleaseChannel == "Unknown" {
+		data.ReleaseChannel = ""
+	}
+
+	clusterBootstrapped, ok := secret.Data["clusterIsBootstrapped"]
+	if ok {
+		data.ClusterBootstrapping = string(clusterBootstrapped) != `"true"`
+	}
+
+	imagesRegistry, ok := secret.Data["imagesRegistry"]
```
Comment (yalosev) [inline deckhouse-controller/pkg/controller/deckhouse-release/controller.go:265]: Shouldn't we read it from the `deckhouse-registry` secret?

Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go]: Fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go (hunk 5 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go]: Fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go (hunk 4 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go:423]: Done

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go (hunk 3 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go:222]: Done

File: deckhouse-controller/pkg/controller/deckhouse-release/cleanup_deckhouse_release.go (hunk 2 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/cleanup_deckhouse_release.go]: Fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/controller.go (hunk 1 above)
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/controller.go]: I haven't found a better way to do this yet

File: deckhouse-controller/pkg/controller/deckhouse-release/check_release.go [hunk 7]
```diff
@@ -28,78 +29,86 @@ import (
...
-		},
-	},
-}, dependency.WithExternalDependencies(checkReleases))
+	for {
+		err := r.checkDeckhouseRelease(ctx)
+		if err != nil {
+			r.logger.Errorf("check Deckhouse release: %s", err)
+		}
+
+		time.Sleep(24 * time.Hour)
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/check_release.go]: Fixed

File: deckhouse-controller/pkg/controller/deckhouse-release/check_release.go [hunk 8]
```diff
@@ -28,78 +29,86 @@ import (
...
 )
 
-const (
-	metricUpdatingFailedGroup = "d8_updating_failed"
-)
+func (r *deckhouseReleaseReconciler) checkDeckhouseReleaseLoop(ctx context.Context) {
+	// Check if Module Manager has been initialized
+	_ = wait.PollUntilContextCancel(ctx, d8utils.SyncedPollPeriod, false,
+		func(context.Context) (bool, error) {
+			return r.moduleManager.AreModulesInited(), nil
```
[the remark this answers is not in the document]
Comment (libmonsoon-dev) [reply deckhouse-controller/pkg/controller/deckhouse-release/check_release.go]: Fixed

**Review by yalosev — CHANGES_REQUESTED** [review]
generates an error
```
{
  "component": "DeckhouseRelease",
  "level": "error",
  "msg": "deploy release: get deployment d8-system/deckhouse: no kind is registered for the type v1.Deployment in scheme \"pkg/runtime/scheme.go:100\"",
  "time": "2024-07-04T18:30:19Z"
}
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @libmonsoon-dev. I've started the workflow for label 'e2e/run/openstack'...

**Review by yalosev — APPROVED** [review]

**Review by miklezzzz — APPROVED** [review]

**Pull request merged**